File: algorithm/python/prob3001-3500/prob3066_medium_超过阈值的最少操作数II.py
import heapq

class Solution:
    def minOperations(self, nums: List[int], k: int) -> int:
        n = len(nums)
        data = nums
        heapq.heapify(data)
        ans = 0
        while len(data) >= 2 and heapq.nsmallest(1, data)[0] < k:
            ans += 1
            x = heapq.heappop(data)
            y = heapq.heappop(data)
            heapq.heappush(data, min(x, y) * 2 + max(x, y))
        return ans

File: algorithm/cpp/prob3001-3500/prob3066_medium_超过阈值的最少操作数II.cpp
#include <queue>

using namespace std;
using ll = long long;

class Solution {
public:
    int minOperations(vector<int>& nums, int k) {
        priority_queue<ll, vector<ll>, greater<ll>> pq(nums.begin(), nums.end());
        int ans = 0;
        while(pq.size() >= 2 && pq.top() < k)
        {
            ans++;
            ll x = pq.top();
            pq.pop();
            ll y = pq.top();
            pq.pop();
            pq.push(min(x, y) * 2 + max(x, y));
        }
        return ans;
    }
};
